File: python/백준/자료구조/스택, 집합/9012 괄호.py
import sys
input = sys.stdin.readline

n = int(input().rstrip())

for _ in range(n):
    string = input().rstrip()
    stack = []
    flag = True  # 올바른 괄호 문자열인지 확인하는 플래그
    for i in range(len(string)):
        if string[i] == '(':
            stack.append('(')
        elif string[i] == ')':
            if not stack or stack[-1] != '(':
                flag = False  # `==` 대신 `=` 사용
                break
            else:
                stack.pop()
    if flag and not stack:  # 스택이 비어 있어야 올바른 문자열
        print("YES")
    else:
        print("NO")

File: cpp/백준/자료구조/스택, 집합/9012 괄호.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    while (n--) {
        string s;
        cin >> s;
        stack<char> st;
        bool flag = true;

        for (char c : s) {
            if (c == '(') {
                st.push('(');
            } else if (c == ')') {
                if (st.empty() || st.top() != '(') {
                    flag = false;
                    break;
                } else {
                    st.pop();
                }
            }
        }

        if (flag && st.empty()) cout << "YES\n";
        else cout << "NO\n";
    }
    return 0;
}
